include/onnxruntime/core/common/logging/logging.h

```diff
@@ -56,6 +58,13 @@ namespace logging {
 
 using Timestamp = std::chrono::time_point<std::chrono::system_clock>;
 
+// NOTE When other compilers support std::chrono::operator<<, update this.
+#if __MAC_OS_X_VERSION_MAX_ALLOWED >= 140400L
```

Comment:
I think the C++20 check is still useful. It just so happens that we only enable C++20 on Mac builds now.

https://github.com/microsoft/onnxruntime/blob/9d06e1bfa469bce92d7aa2ebfa36c7a9714ba119/cmake/CMakeLists.txt#L42-L47

How about something like this?

```suggestion
// Xcode added support for C++20's std::chrono::operator<< in version <the version>.
#if __cplusplus >= 202002L && (!defined(__MAC_OS_X_VERSION_MAX_ALLOWED) || __MAC_OS_X_VERSION_MAX_ALLOWED >= 140400L)
```

Reply:
We did not test other compilers yet. 
https://en.cppreference.com/w/cpp/compiler_support/20
![image](https://github.com/microsoft/onnxruntime/assets/30328909/5318bb50-b0e3-4a08-b1f5-e31d3eaa1196)

I suggest to use the following for now:
```
#if __cplusplus >= 202002L &&  defined(__APPLE__)
```

Note that Xcode 14+ uses c++ 20 as default.

Reply:
The issue is that Xcode 14.2 doesn't implement this, but later versions of Xcode do (not sure which exact version it got added in). So we'd need a check based on the Xcode version to be able to build for both 14.2 (the CI build version) and a more current version.

Reply:
The other compiler might have similar issue. From the above table, I think gcc 14 or MSVC 19.29 shall be able to build this, but early version is not guaranteed since only partial implemented.

Reply:
Ah, good point. We'll have to deal with those issues once we enable C++20 for other compilers. Perhaps a comment about needing to test with other compilers later is good enough for this PR.

Reply:
Thank you for the code suggestions. I think it is a much cleaner solution. I can add a comment about the C++20 tests for other compilers.

I am on Xcode 14.4, and we know that 14.2 does not implement `std::chrono::operator<<` from the CI build. According to [Xcode wiki](https://en.wikipedia.org/wiki/Xcode) and [xcodereleases.com](https://xcodereleases.com/), it seems like there was no version 14.3. 

An alternative that we can consider is using the [C-style date-time library](https://en.cppreference.com/w/cpp/chrono/c/time). This would require a more significant rewrite, but should be consistent between compiler versions since it is specified by POSIX (Personally don't like this option).

Reply:
> According to [Xcode wiki](https://en.wikipedia.org/wiki/Xcode) and [xcodereleases.com](https://xcodereleases.com/), it seems like there was no version 14.3.

Thanks for the links. It looks like there is Xcode 14.3 and then Xcode 15, but no Xcode 14.4? Mind verifying if you're running Xcode 15?

We were running builds on Xcode 14.3.1 until recently (downgraded build agent images) and I don't recall seeing this issue before.

> An alternative that we can consider is using the [C-style date-time library](https://en.cppreference.com/w/cpp/chrono/c/time). This would require a more significant rewrite, but should be consistent between compiler versions since it is specified by POSIX (Personally don't like this option).

I think we can just use `std::chrono::operator<<` once we fully move to C++20 and all the compilers we target support it.

Reply:
> Thanks for the links. It looks like there is Xcode 14.3 and then Xcode 15, but no Xcode 14.4? Mind verifying if you're running Xcode 15? 

Sorry I was not specific. The `__MAC_OS_X_VERSION_MAX_ALLOWED` macro corresponds to the SDK version. If we compare against the Xcode 14.2 version, the SDK is version 13.1.

I am running Xcode 15. Here is the output of `clang --version`. 
```
❯ clang --version
Apple clang version 15.0.0 (clang-1500.3.9.4)
Target: arm64-apple-darwin23.4.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
❯ xcode-select --version
xcode-select version 2406.
```

Reply:
Ah, got it. Thanks for clarifying. Maybe the comment can also mention that 14.4 is the SDK version. Otherwise, it looks good.